extensions-core/hdfs-storage/src/main/java/org/apache/druid/inputsource/hdfs/HdfsInputSource.java

```diff
@@ -0,0 +1,238 @@
+ * under the License.
+ */
+
+package org.apache.druid.inputsource.hdfs;
+
+import com.fasterxml.jackson.annotation.JacksonInject;
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import org.apache.druid.data.input.AbstractInputSource;
+import org.apache.druid.data.input.InputFormat;
+import org.apache.druid.data.input.InputRowSchema;
+import org.apache.druid.data.input.InputSourceReader;
+import org.apache.druid.data.input.InputSplit;
+import org.apache.druid.data.input.SplitHintSpec;
+import org.apache.druid.data.input.impl.InputEntityIteratingReader;
+import org.apache.druid.data.input.impl.SplittableInputSource;
+import org.apache.druid.java.util.common.IAE;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.RemoteIterator;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Spliterator;
+import java.util.Spliterators;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+public class HdfsInputSource extends AbstractInputSource implements SplittableInputSource<Path>
+{
+  private static final String PROP_PATHS = "paths";
+
+  private final List<String> inputPaths;
+  private final Configuration configuration;
+
+  @JsonCreator
+  public HdfsInputSource(
+      @JsonProperty(PROP_PATHS) Object inputPaths,
+      @JacksonInject Configuration configuration
+  )
+  {
+    // Note: Currently commas or globs in InputPaths are not supported
+    this.inputPaths = coerceInputPathsToList(inputPaths);
+
+    this.configuration = configuration;
+  }
+
+  private static List<String> coerceInputPathsToList(Object inputPaths)
+  {
+    final List<String> paths;
+
+    if (inputPaths instanceof String) {
+      paths = Collections.singletonList((String) inputPaths);
+    } else if (inputPaths instanceof List && ((List<?>) inputPaths).stream().allMatch(x -> x instanceof String)) {
+      paths = ((List<?>) inputPaths).stream().map(x -> (String) x).collect(Collectors.toList());
+    } else {
+      throw new IAE("'%s' must be a string or an array of strings", PROP_PATHS);
+    }
+
+    return paths;
+  }
+
+
+  @JsonProperty(PROP_PATHS)
+  private List<String> getInputPaths()
+  {
+    return inputPaths;
+  }
+
+  @Override
+  protected InputSourceReader formattableReader(
+      InputRowSchema inputRowSchema,
+      @Nullable InputFormat inputFormat,
+      @Nullable File temporaryDirectory
+  )
+  {
+    return new InputEntityIteratingReader(
+        inputRowSchema,
+        inputFormat,
+        createSplits(null, null).map(split -> new HdfsSource(configuration, split.get())),
```

Comment:
Passing `null` to a non-`@Nullable` method looks strange, but I guess it's fine in the context of this class. Maybe annotate `inputFormat` as `@Nullable` in this override only, even if it's not in the interface. Or have that method, and this call site, call a common helper method.

Reply:
Hmm, I would recommend passing `inputFormat` properly instead since it's easy. It should be used properly once we support splitting a large files into smaller splits.

Reply:
I've removed the `@Nullable` for `inputFormat` and passed it to `createSplits()`